staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/complete.go

```diff
@@ -66,17 +66,29 @@ func validateNestedValueValidationCompleteness(v *NestedValueValidation, s *Stru
 
 	allErrs := field.ErrorList{}
 
-	allErrs = append(allErrs, validateValueValidationCompleteness(&v.ValueValidation, s, sPath, vPath)...)
-	allErrs = append(allErrs, validateNestedValueValidationCompleteness(v.Items, s.Items, sPath.Child("items"), vPath.Child("items"))...)
+	allErrs = append(allErrs, validateValueValidationCompleteness(&v.ValueValidation, s, sPath, vPath, opts)...)
+	allErrs = append(allErrs, validateNestedValueValidationCompleteness(v.Items, s.Items, sPath.Child("items"), vPath.Child("items"), opts)...)
+
+	var additionalPropertiesSchema *Structural
+	if s.AdditionalProperties != nil && s.AdditionalProperties.Structural != nil {
+		additionalPropertiesSchema = s.AdditionalProperties.Structural
+	}
+
 	for k, vFld := range v.Properties {
 		if sFld, ok := s.Properties[k]; !ok {
-			allErrs = append(allErrs, field.Required(sPath.Child("properties").Key(k), fmt.Sprintf("because it is defined in %s", vPath.Child("properties").Key(k))))
+			if additionalPropertiesSchema == nil || !opts.AllowValidationPropertiesWithAdditionalProperties {
+				allErrs = append(allErrs, field.Required(sPath.Child("properties").Key(k), fmt.Sprintf("because it is defined in %s", vPath.Child("properties").Key(k))))
+			} else {
+				allErrs = append(allErrs, validateNestedValueValidationCompleteness(&vFld, additionalPropertiesSchema, sPath.Child("additionalProperties"), vPath.Child("properties").Key(k), opts)...)
+			}
 		} else {
-			allErrs = append(allErrs, validateNestedValueValidationCompleteness(&vFld, &sFld, sPath.Child("properties").Key(k), vPath.Child("properties").Key(k))...)
+			allErrs = append(allErrs, validateNestedValueValidationCompleteness(&vFld, &sFld, sPath.Child("properties").Key(k), vPath.Child("properties").Key(k), opts)...)
 		}
 	}
 
-	// don't check additionalProperties as this is not allowed (and checked during validation)
+	if v.AdditionalProperties != nil && opts.AllowNestedAdditionalProperties {
+		allErrs = append(allErrs, validateNestedValueValidationCompleteness(v.AdditionalProperties, s.AdditionalProperties.Structural, sPath.Child("additionalProperties"), vPath.Child("additionalProperties"), opts)...)
+	}
```

Comment:
wondering, we have recursions for:
1. property in s, property in v
2. additionalProperty in s, property in v
3. additionalProperty in s, additionalProperty in v

What about the last case? property in s, additionalProperty in v

```yaml
properties:
  a:
    type: object
    properties:
      x:
        type: string
allOf:
  addtionalProperties:
    properties:
      y:
        pattern: .*
```

Is this schema complete? `a.y` is not structurally specified, but should?

Reply:
This schema is incomplete, and caught by the code today. 

I think the following schema should be seen as complete (today it is seen as incomplete):

```
properties:
  a:
    type: object
    properties:
      x:
        type: string
allOf:
  addtionalProperties:
    properties:
      x:
        pattern: .*
```

Reply:
Right. And are those detected as incomplete?

Reply:
Having test cases for all of these would make me feel better, basically testing your updated go doc:

> additionalProperties validations must have additionalProperties defined in the structural portion of the schema corresponding to that node

Reply:
Added